src/native/iterator.cc

```diff
@@ -0,0 +1,331 @@
+          DescendRight();
+          break;
+        } else {
+          InsertRightChild(end_offset);
+          DescendRight();
+          MarkLeft(id, start_offset, end_offset);
+          return node;
+        }
+    }
+  }
+}
+
+Node* Iterator::InsertSpliceBoundary(const Point &offset, bool is_insertion_end) {
+  Reset();
+
+  while (true) {
+    int comparison = offset.Compare(node_offset);
+    if (comparison == 0 && !is_insertion_end) {
+      return node;
+    } else if (comparison < 0) {
+      if (node->left) {
+        DescendLeft();
+      } else {
+        InsertLeftChild(offset);
+        return node->left;
+      }
+    } else { // comparison > 0
+      if (node->right) {
+        DescendRight();
+      } else {
+        InsertRightChild(offset);
+        return node->right;
+      }
+    }
+  }
+}
+
+void Iterator::FindIntersecting(const Point &start, const Point &end, std::unordered_set<MarkerId> *result) {
+  Reset();
+
+  if (!node) return;
+
+  while (true) {
+    if (start < node_offset) {
+      if (node->left) {
+        CheckIntersection(start, end, result);
+        DescendLeft();
+      } else {
+        break;
+      }
+    } else {
+      if (node->right) {
+        CheckIntersection(start, end, result);
+        DescendRight();
+      } else {
+        break;
+      }
+    }
+  }
+
+  do {
+    CheckIntersection(start, end, result);
+    MoveToSuccessor();
+  } while (node && node_offset <= end);
+}
+
+void Iterator::FindContainedIn(const Point &start, const Point &end, std::unordered_set<MarkerId> *result) {
+  Reset();
+
+  if (!node) return;
+
+  SeekToFirstNodeGreaterThanOrEqualTo(start);
+
+  unordered_set<MarkerId> started;
+  while (node && node_offset <= end) {
+    started.insert(node->start_marker_ids.begin(), node->start_marker_ids.end());
+    for (const MarkerId &id : node->end_marker_ids) {
+      if (started.count(id) > 0) result->insert(id);
+    }
+    MoveToSuccessor();
+  }
+}
+
+void Iterator::FindStartingIn(const Point &start, const Point &end, std::unordered_set<MarkerId> *result) {
+  Reset();
+
+  if (!node) return;
+
+  SeekToFirstNodeGreaterThanOrEqualTo(start);
+
+  while (node && node_offset <= end) {
+    result->insert(node->start_marker_ids.begin(), node->start_marker_ids.end());
+    MoveToSuccessor();
+  }
+}
+
+void Iterator::FindEndingIn(const Point &start, const Point &end, std::unordered_set<MarkerId> *result) {
+  Reset();
+
+  if (!node) return;
+
+  SeekToFirstNodeGreaterThanOrEqualTo(start);
+
+  while (node && node_offset <= end) {
+    result->insert(node->end_marker_ids.begin(), node->end_marker_ids.end());
+    MoveToSuccessor();
+  }
+}
+
+unordered_map<MarkerId, Range> Iterator::Dump() {
+  Reset();
+
+  unordered_map<MarkerId, Range> snapshot;
+
+  if (!node) return snapshot;
+
+  while (node && node->left) DescendLeft();
+
+  while (node) {
+    for (const MarkerId &id : node->start_marker_ids) {
+      Range range;
+      range.start = node_offset;
```

Comment:
I think it would be a little clearer to explicitly initialize the range's `end` to zero, like `Range range{node_offset, Point(0, 0)};`.